### PR TITLE
Refactor Grades Today Widget to new architecture

### PR DESCRIPTION
**Pull request by shichongrui**
refs: MBL-12172
affects: Student
release note: none

### CODE REVIEW

**Comment by inst-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  This pull request will not generate a release note.
  </td>
    </tr>
  </tbody>
</table>


#### Affected Apps: Student

[MBL-12172](https://instructure.atlassian.net/browse/MBL-12172)

Coverage | New % | Master % | Delta
-------- | ----- | -------- | -----
**Core** | 93.99% | 93.91% | 0.08%
**Student** | 93.01% | 93.01% | 0%
**Teacher** | 100% | 100% | 0%
**React Native** | 92.53% | 92.53% | 0%
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: Canvas/GradesWidget/GradesTodayWidgetPresenter.swift [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        })
+    }()
+
+    init(env: AppEnvironment = .shared, view: GradesTodayWidgetViewController) {
+        self.env = env
+        self.view = view
+    }
+
+    func viewIsReady() {
+        courses.refresh(force: true)
```
Comment (bluwave) [inline Canvas/GradesWidget/GradesTodayWidgetPresenter.swift]: do we want these to do a force refresh every time?

File: Canvas/GradesWidget/GradesTodayWidgetViewController.swift [hunk 2]
```diff
@@ -80,135 +64,55 @@ func isDeviceLocked() -> Bool {
...
-        tableView.tableFooterView = UIView(frame: .zero)
 
-        // Cells
         let nib = UINib(nibName: String(describing: GradeWidgetCell.self), bundle: nil)
         tableView.register(nib, forCellReuseIdentifier: "cell")
 
-        return tableView
-    }()
-
-    override func didReceiveMemoryWarning() {
```
Comment (bluwave) [inline Canvas/GradesWidget/GradesTodayWidgetViewController.swift]: we may want to leave this `didReceiveMemoryWarning` method.  The widgets don't get as much memory as apps do.  I think this was pivotal to finding a bug if i remember right, @armstrongnate can you confirm how you stress tested the widget in the past?

Comment (armstrongnate) [reply Canvas/GradesWidget/GradesTodayWidgetViewController.swift]: That's right. This warning is especially relevant for today widgets.

File: Canvas/GradesWidget/GradesTodayWidgetViewController.swift [hunk 3]
```diff
@@ -277,19 +163,62 @@ extension GradesTodayWidgetViewController: UITableViewDataSource {
...
+    func displayGrade(for course: Core.Course) -> String {
+        guard let enrollments = course.enrollments, let enrollment = enrollments.filter({ $0.role == .student }).first else {
+            return ""
+        }
+        
+        let grade = enrollment.multipleGradingPeriodsEnabled
+            ? enrollment.currentPeriodComputedCurrentGrade
+            : enrollment.computedCurrentGrade
+
+        let formatter = NumberFormatter()
```
Comment (bluwave) [inline Canvas/GradesWidget/GradesTodayWidgetViewController.swift]: It's usually best practice not to init new formatters in `cellForRowAtIndexPath`.   Have you considered making this formatter a class var to help table view methods be more efficient?

File: Core/CoreTests/Database/Models/EnrollmentTests.swift [hunk 4]
```diff
@@ -30,6 +30,23 @@ class EnrollmentTests: XCTestCase {
         XCTAssertEqual(EnrollmentRole.student.rawValue, "StudentEnrollment")
     }
 
+    func testUpdateFromAPI() {
```
Comment (bluwave) [inline Core/CoreTests/Database/Models/EnrollmentTests.swift]: great job

**Comment by bluwave** [comment]
When i clicked on a course grade cell in the widget, it took me to the dashboard instead of the course grades page

File: Core/Core/Database/CoreData/Models/Course.swift [hunk 5]
```diff
@@ -68,6 +68,45 @@ extension Course {
...
+        var score = enrollment.computedCurrentScore
+
+        if enrollment.multipleGradingPeriodsEnabled && enrollment.currentGradingPeriodID != nil {
+            grade = enrollment.currentPeriodComputedCurrentGrade
+            score = enrollment.currentPeriodComputedCurrentScore
+        } else if enrollment.multipleGradingPeriodsEnabled && enrollment.totalsForAllGradingPeriodsOption {
+            grade = enrollment.computedFinalGrade
+            score = enrollment.computedFinalScore
+        } else if enrollment.multipleGradingPeriodsEnabled && enrollment.totalsForAllGradingPeriodsOption == false {
+            return "N/A"
```
Comment (armstrongnate) [inline Core/Core/Database/CoreData/Models/Course.swift]: I _believe_ this needs to be localized using `NSLocalizedString`.

File: Core/Core/Database/CoreData/Models/Course.swift [hunk 6]
```diff
@@ -68,6 +68,45 @@ extension Course {
...
+        let formatter = NumberFormatter()
+        formatter.numberStyle = .percent
+        formatter.decimalSeparator = "."
+        formatter.multiplier = 1
+        formatter.maximumFractionDigits = 3
+        formatter.roundingMode = .down
+        return formatter
+    }()
+
+    public var widgetDisplayGrade: String {
```
Comment (armstrongnate) [inline Core/Core/Database/CoreData/Models/Course.swift]: Do you think we can call this `displayGrade`? Ideally the widget will show the same thing as the course cards and everywhere else, right?

File: Core/CoreTests/Database/Models/CourseTests.swift [hunk 7]
```diff
@@ -84,4 +84,90 @@ class CourseTests: CoreTestCase {
         XCTAssertNotNil(resultEnrollment)
         XCTAssertEqual(resultEnrollment?.canvasContextID, "course_1")
     }
+
```
Comment (armstrongnate) [inline Core/CoreTests/Database/Models/CourseTests.swift]: Great tests! WOOHOO!

**Pull request merged**